### ISSUE TITLE
Gradient Checkpointing does not update BN Layers Weights & Biases

### ISSUE THREAD

**Issue by ai-padawan** [KeremTurgutlu/self_supervised#63]
When using the CheckpointSequential module to do gradient checkpointing, the BN layers weights do not get updated.

You can test this on the standard template in the repo, keep the same dataset, batch size, epoch, etc...
1) If you trained without CheckpointSequential, sample output is:
----------------encoder.4.0.bn1.weight--------------------------
tensor([1.0012, 1.0022, 0.9971, 0.9942, 0.9960, 0.9988, 0.9976, 1.0038, 1.0001,
        1.0055, 0.9962, 0.9926, 1.0015, 1.0000, 1.0018, 1.0036, 0.9987, 1.0021,
        0.9996, 1.0071, 1.0000, 0.9942, 1.0038, 1.0016, 1.0002, 0.9970, 1.0007,
        0.9935, 0.9930, 0.9913, 1.0025, 1.0045, 1.0044, 0.9915, 0.9999, 1.0043,
        0.9995, 0.9987, 0.9985, 0.9953, 0.9979, 1.0013, 1.0019, 1.0001, 0.9955,
        1.0036, 0.9937, 1.0079, 1.0100, 0.9989, 1.0008, 0.9986, 0.9984, 1.0034,
        1.0037, 0.9960, 1.0040, 0.9972, 1.0049, 1.0057, 1.0009, 0.9961, 1.0024,
        1.0066])
----------------encoder.4.0.bn1.bias--------------------------
tensor([ 1.0627e-03, -1.1718e-03, -7.5741e-03, -3.5472e-03, -6.9174e-03,
         4.5058e-03, -2.8276e-03, -2.1415e-03,  6.6311e-03,  2.1298e-03,
        -3.0902e-03, -1.7534e-04,  3.0238e-03, -1.2597e-03, -1.0779e-03,
         4.0659e-03,  1.4349e-03, -1.5296e-03,  6.4844e-03,  1.0662e-02,
        -3.6205e-03, -5.9499e-03,  2.1282e-03,  1.3002e-03, -7.1395e-04,
        -3.9172e-03,  4.8558e-03, -4.1608e-03, -2.7316e-03, -6.8490e-03,
         4.2301e-03,  8.8376e-04, -1.9853e-03, -3.5704e-03,  2.1841e-03,
        -8.0314e-04, -1.7476e-03,  9.0136e-04,  2.4802e-04, -5.2012e-03,
        -9.9274e-04,  2.8108e-04,  6.7686e-05, -3.7550e-03, -3.8663e-03,
         1.5729e-03,  1.0614e-03,  1.2925e-02,  2.7174e-03,  3.6385e-03,
        -2.9763e-03,  1.7430e-03,  3.6633e-03,  4.3344e-03,  3.9695e-03,
        -1.4936e-03,  1.7176e-03, -6.0875e-03,  5.6964e-03,  5.0546e-03,
         5.7726e-03, -2.1435e-03, -6.2057e-03,  3.2758e-03])
2) If you trained with Checkpoint Sequential, sample output:
----------------encoder.fastai_model.4.0.bn1.weight--------------------------
tensor([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
----------------encoder.fastai_model.4.0.bn1.bias--------------------------
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

I've verified this using ResNet50 backbone. Even after training with 100 epoch with Checkpoint Sequential the weight/bias remains exactly at 1/0. Tested on both SimCLR and Barlow losses. This affects downstream performance.

The only modification to my code is replacing the encoder with a checkpointed encoder:
`encoder = CheckpointSequential(encoder, checkpoint_nchunks=10)`

**Comment by KeremTurgutlu** [comment]
Thanks for letting me know, I will take a look at it. Have you seen any similar issues regarding this in PyTorch? We are using the built-in PyTorch utils for gradient checkpointing.

**Comment by KeremTurgutlu** [comment]
This happens in the case of fastai, and not for timm so to unblock yourself you can use timm resnet models with `CheckpointResNet`.

The following shows True for the BN params that are not updated during training if we use fastai resnet encoder:
```
fastai_model.4.0.bn1.weight True
fastai_model.4.0.bn1.bias True
fastai_model.4.0.bn2.weight True
fastai_model.4.0.bn2.bias True
fastai_model.4.1.bn1.weight True
fastai_model.4.1.bn1.bias True
fastai_model.4.1.bn2.weight True
fastai_model.4.1.bn2.bias True
fastai_model.5.0.bn1.weight False
fastai_model.5.0.bn1.bias False
fastai_model.5.0.bn2.weight False
fastai_model.5.0.bn2.bias False
fastai_model.5.1.bn1.weight False
fastai_model.5.1.bn1.bias False
fastai_model.5.1.bn2.weight False
fastai_model.5.1.bn2.bias False
fastai_model.6.0.bn1.weight False
fastai_model.6.0.bn1.bias False
fastai_model.6.0.bn2.weight False
fastai_model.6.0.bn2.bias False
fastai_model.6.1.bn1.weight False
fastai_model.6.1.bn1.bias False
fastai_model.6.1.bn2.weight False
fastai_model.6.1.bn2.bias False
fastai_model.7.0.bn1.weight False
fastai_model.7.0.bn1.bias False
fastai_model.7.0.bn2.weight False
fastai_model.7.0.bn2.bias False
fastai_model.7.1.bn1.weight False
fastai_model.7.1.bn1.bias False
fastai_model.7.1.bn2.weight False
fastai_model.7.1.bn2.bias False
```
